### ISSUE TITLE
3D convolutions unnaturally slow on CPU

### ISSUE THREAD

**Issue by undertherain** [tensorflow/tensorflow#7610]
3D convolutions on CPU seem unnaturally slow. 
I can't use GPU due to memory limit, so I'm looking at CPU execution. 

2D convolutions in TensorFlow seem to be well-optimized, all CPU cores are used, performance is just few times below GPU.

With 3D convolutions - the difference is orders of magnitude. 

Also, I've compared with Theano. Theano 3D convolutions run on single core but still are 10 times faster.

Strangely, TF uses all cores on CPU with 3Dconv,  so there must be some bug or extreme inefficiency in implementation.

With the same small test model (just couple of 3d conv layers) I get 1 second epoch time on GPU (both TF and Theano, theano just a bit faster), 5 seconds on CPU Theano single threaded, and 50 seconds with (seemingly) multi-threded TF

### Environment info
I've tried different TF versions from 0.12 to 1.00, installed from pip. Latest one from https://storage.googleapis.com/tensorflow/linux/gpu/tensorflow_gpu-1.0.0-cp35-cp35m-linux_x86_64.whl

The OS is CentOS Linux release 7.2.1511 (Core) , kernel 3.10.0-327.18.2.el7.x86_64

I'm using Keras back-end, so I can't be 100% sure if the problem is not in the way Keras translates Convolution3D call into TF primitives, but it does not seem likely. 

cheers
Alex

**Comment by yaroslavvb** [comment]
cc @daeyun who added conv3d_transpose

(ps: who maintains core/kernels/conv_3d.h  ?)

**Comment by vrv** [comment]
@mjanusz might know, I think he implemented / uses Conv3D, possibly even on CPU.

**Comment by mjanusz** [comment]
Eigen's conv3d it not optimized to the same level as conv2d, but the speed you are observing seems to be particularly bad. I will take a look at this.

**Comment by urimerhav** [comment]
I can affirm this. I've experienced an unnatural 100X factor between a latest mac and a gpu machine. On other tasks I got used to seeing 5-20X at most with 2d convolutions and matrix multiplications.

**Comment by girving** [comment]
@mjanusz Did you get a chance to look at this?

**Comment by yonidonner** [comment]
Building from sources (MKL enabled) seems to help a lot with this issue. Compared to the pip binary, almost 10x faster.

**Comment by mjanusz** [comment]
@girving Yes. The commit a999474 fixes a regression that should provide a 3x-5x speedup, depending on the exact configuration. The next step to improve performance is to write a contraction mapper in eigen_cuboid_convolutions.h, but I haven't had the chance to do that yet.

**Comment by mmuneebs** [comment]
Observing the same problem here on macOS with CPU. Unnaturally slow and extreme memory requirements (50-100x more compared to conv2d).

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
A member of the TensorFlow organization has replied after the stat:awaiting tensorflower label was applied.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tatianashp** [comment]
@mjanusz Are you still planning on writing a contraction mapper in eigen_cuboid_convolutions.h? I'll keep this issue open for tracking purposes in case you do.

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tatianashp** [comment]
Automatically closing due to lack of recent activity. Please update the issue when new information becomes available, and we will reopen the issue. Thanks!

**Comment by josuerocha** [comment]
I am also using 3D convolutions and they are unnaturally slow on CPU. Is there any plan to optimize 3D convolutions?

**Comment by innat** [comment]
Any update on this?

This issue opened in **2017**
The cause is same in **2023** with TF 2.12

# Ps. lesson learned: don't close the issue if such critical issue isn't solved yet.

I was porting a torch model to tf on cpu machine that uses conv3D. I have just lost coulple of hours to finding mismatch in performance due to such issue. A proverb goes, life is short, switch to torch. 😅
